### PR TITLE
Add satoshi denomination

### PR DESCRIPTION
**Pull request by AbelLykens**
* Related: https://github.com/bitcoin-wallet/bitcoin-wallet/pull/594#issuecomment-552184448
* Add satoshi denomination for bitcoinj
* Regular definition (1/10000000th bitcoin)
* Display as sat (so this does not include the bitcoin symbol)

### CODE REVIEW

File: core/src/main/java/org/bitcoinj/utils/MonetaryFormat.java [hunk 1]
```diff
@@ -60,12 +62,16 @@
...
+    /** Currency symbol for base 1/100000000 Bitcoin. */
+    public static final String CODE_SAT = "sat";
     /** Currency symbol for base 1 Bitcoin. */
     public static final String SYMBOL_BTC = "\u20bf";
     /** Currency symbol for base 1/1000 Bitcoin. */
     public static final String SYMBOL_MBTC = "m" + SYMBOL_BTC;
     /** Currency symbol for base 1/1000000 Bitcoin. */
     public static final String SYMBOL_UBTC = "µ" + SYMBOL_BTC;
+    /** Currency symbol for base 1/100000000 Bitcoin (satoshi). */
+    public static final String SYMBOL_UBTC = "sat";
```
Comment (HashEngineering) [inline core/src/main/java/org/bitcoinj/utils/MonetaryFormat.java]: This is a duplicate symbol SYMBOL_UBTC and as a result the Travis CI tests fail.

**Comment by HashEngineering** [comment]
You should also add some unit tests for this new monetary unit.

File: core/src/main/java/org/bitcoinj/utils/MonetaryFormat.java (hunk 1 above)
Comment (AbelLykens) [reply core/src/main/java/org/bitcoinj/utils/MonetaryFormat.java]: renamed

**Comment by AbelLykens** [comment]
Added a simple test line, but if someone with experience could look at it, please do

File: core/src/main/java/org/bitcoinj/utils/MonetaryFormat.java [hunk 2]
```diff
@@ -60,12 +62,16 @@
...
     public static final String CODE_UBTC = "µBTC";
+    /** Currency symbol for base 1/100000000 Bitcoin. */
+    public static final String CODE_SAT = "sat";
     /** Currency symbol for base 1 Bitcoin. */
     public static final String SYMBOL_BTC = "\u20bf";
     /** Currency symbol for base 1/1000 Bitcoin. */
     public static final String SYMBOL_MBTC = "m" + SYMBOL_BTC;
     /** Currency symbol for base 1/1000000 Bitcoin. */
     public static final String SYMBOL_UBTC = "µ" + SYMBOL_BTC;
+    /** Currency symbol for base 1/100000000 Bitcoin (satoshi). */
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/utils/MonetaryFormat.java]: I don't think "sat" is a symbol. We probably don't have a symbol for the satoshi unit already. I suggest to remove this constant – you already added "sat" as a currency code.

Comment (AbelLykens) [reply core/src/main/java/org/bitcoinj/utils/MonetaryFormat.java]: Thanks. Removed.

**Comment by AbelLykens** [comment]
Added to constructor. Thanks for your help so far @schildbach , I learn a lot from this ;). 
I think the new test fails the build [?]. Wonder what I did wrong here.

```
> Task :bitcoinj-core:test
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by net.sf.cglib.core.ReflectUtils$2 (file:/home/travis/.gradle/caches/modules-2/files-2.1/cglib/cglib-nodep/2.2.2/d456bb230c70c0b95c76fb28e429d42f275941/cglib-nodep-2.2.2.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of net.sf.cglib.core.ReflectUtils$2
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
org.bitcoinj.utils.MonetaryFormatTest > standardCodes FAILED
    java.util.DuplicateFormatFlagsException: Flags = '0'
        at java.base/java.util.Formatter$Flags.parse(Formatter.java:4605)
        at java.base/java.util.Formatter$FormatSpecifier.flags(Formatter.java:2800)
        at java.base/java.util.Formatter$FormatSpecifier.<init>(Formatter.java:2854)
        at java.base/java.util.Formatter.parse(Formatter.java:2713)
        at java.base/java.util.Formatter.format(Formatter.java:2655)
        at java.base/java.util.Formatter.format(Formatter.java:2609)
        at java.base/java.lang.String.format(String.java:2938)
        at org.bitcoinj.utils.MonetaryFormat.format(MonetaryFormat.java:366)
        at org.bitcoinj.utils.MonetaryFormatTest.standardCodes(MonetaryFormatTest.java:217)


```

File: core/src/test/java/org/bitcoinj/utils/MonetaryFormatTest.java [hunk 3]
```diff
@@ -214,7 +214,6 @@ public void standardCodes() throws Exception {
         assertEquals("BTC 0.00", MonetaryFormat.BTC.format(Coin.ZERO).toString());
         assertEquals("mBTC 0.00", MonetaryFormat.MBTC.format(Coin.ZERO).toString());
         assertEquals("µBTC 0", MonetaryFormat.UBTC.format(Coin.ZERO).toString());
-        assertEquals("sat 0", MonetaryFormat.SAT.format(Coin.ZERO).toString());
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/utils/MonetaryFormatTest.java]: Please keep the test in the PR. We need tests.

**Comment by AbelLykens** [comment]
Hope someone who understands [the] tests can look at it. Not sure what to add...

**Comment by schildbach** [comment]
The test is fine (I think).

It's the implementation that lacks. You need to add your denomination to the codes array (`this.codes[8]`) in the second `MonetaryFormat` constructor.

**Review by monstro805 — APPROVED** [review]

**Review by monstro805 — APPROVED** [review]

**Comment by schildbach** [comment]
Support for satoshi denomination was added via 6a9893c3df2662e8ad099b1a1a1ffc49c3a42320.

**Pull request closed without merging**